Step04/app/src/main/java/com/kwoolytech/step04/MainActivity.java

```diff
@@ -0,0 +1,168 @@
+package com.kwoolytech.step04;
+
+import android.graphics.Bitmap;
+import android.support.design.widget.NavigationView;
+import android.support.v4.view.GravityCompat;
+import android.support.v4.widget.DrawerLayout;
+import android.support.v7.app.ActionBarDrawerToggle;
+import android.support.v7.app.AppCompatActivity;
+import android.os.Bundle;
+import android.support.v7.widget.Toolbar;
+import android.util.Log;
+import android.view.Menu;
+import android.view.MenuItem;
+import android.view.View;
+import android.widget.Button;
+import android.widget.ListView;
+
+public class MainActivity extends AppCompatActivity implements NavigationView.OnNavigationItemSelectedListener {
+    private HansWeatherDataModel   dataModel;
+    private KwoolyHttp             httpClient;
+    private KwoolyHttpCallback     httpClientCallback;
+
+    private HansWeatherViewAdapter listViewWeatherAdapter;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_main);
+
+        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
+        setSupportActionBar(toolbar);
+
+        DrawerLayout drawer = (DrawerLayout)findViewById(R.id.drawer_layout);
+        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(this, drawer, toolbar,
+                R.string.navigation_drawer_open, R.string.navigation_drawer_close);
+
+        drawer.setDrawerListener(toggle);
+        toggle.syncState();
+
+        NavigationView navigationView = (NavigationView)findViewById(R.id.nav_view);
+        navigationView.setNavigationItemSelectedListener(this);
+
+        dataModel = new HansWeatherDataModel();
+        httpClient  = new KwoolyHttp(MainActivity.this);
+
+        initializeUi();
+        initializeKwoolyHttpCallbackFunction();
+
+        queryWeatherJsonData();
+
+        return;
+    }
+
+    private void initializeUi() {
+        listViewWeatherAdapter = new HansWeatherViewAdapter(MainActivity.this, dataModel,
+                                         R.layout.listviewitem_weather);
+        ((ListView)findViewById(R.id.listViewWeather)).setAdapter(listViewWeatherAdapter);
+
+        findViewById(R.id.buttonRefresh).setOnClickListener(new Button.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                queryWeatherJsonData();
+            }
+        });
+        return;
+    }
+
+    private void initializeKwoolyHttpCallbackFunction() {
+        httpClientCallback = new KwoolyHttpCallback() {
+            @Override
+            public void onWeatherJsonDataReceived(String result) {
+                try {
+                    dataModel.setWeatherJsonData(result);
+                    queryWeatherBitmapData();
```

Comment:
@kwoolytech I think it is better to show weather information in a text format first to user and then update the weather image after receiving it in a row so user doesn't wait for two steps to see the list item.